fastcache.go

```diff
@@ -121,6 +122,26 @@ func (c *Cache) UpdateStats(s *Stats) {
 	}
 }
 
+// Keys retrieves all cached keys matching regex pattern
+func (c *Cache) Keys(pattern string) (keys [][]byte, err error) {
+	r, err := regexp.Compile(pattern)
+	if err != nil {
+		return
+	}
+
+	for _, b := range c.buckets {
+		for _, chunk := range b.chunks {
+			if len(chunk) > 0 {
+				if key := chunk[4 : 4+chunk[1]]; r.Match(key) {
```

Comment:
This is completely wrong code. It may check for a single key per chunk if the key size is less than 255 bytes. In other cases it may do anything :)

Reply:
Apparently, I misunderstood how items are stored. I thought each chunk has a header which indicates where the cache key is positioned inside the chunk.

Reply:
Each chunk contains marshaled `(key, value)` pairs which go one after the other. The first two bytes contain key lengths, the next two bytes - value length followed by key contents and then by value contents. In general it is impossible to decode all the `(key, value)` pairs from chunks. The decoding must go from `bucket.m`. And don't forget to hold `bucket.mu` lock while accessing `bucket` contents.

Reply:
I've pushed the visitor implementation. Indeed, the number of allocations is at the minimum.

The cost is the "f" function can change cached data. If we'd use copies of key and value to call f, there would be many allocations, but the garbage collector would run after each f call, wouldn't it?